GPT/gpt-function-calling.py

```diff
@@ -0,0 +1,185 @@
+                            "str": {
+                                "type": "string",
+                                "description": "The text to display",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            },
+            {
+                "type": "function",
+                "function": {
+                    "name": "search_for_command",
+                    "description": "search_for_command(str: string) - this searches for a command in the VSCode command palette. If I ask you to do something, please use this command to search for an appropriate command.",
+                    "parameters": {
+                        "type": "object",
+                        "properties": {
+                            "str": {
+                                "type": "string",
+                                "description": "The command to search for",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            }
+        ],
+        "max_tokens": 2024,
+        "temperature": 0.6,
+        "n": 1,
+        "stop": None,
+        "model": settings.get("user.openai_model"),
+    }
+
+
+    response = requests.post(url, headers=headers, data=json.dumps(data))
+
+    if response.status_code == 200:
+        notify("GPT Task Completed")
+        print(response.json())
+
+        process_function_calls(insert_response, response)
+
+        content = (response.json()["choices"][0]["message"]["content"] or "").strip()
+        if len(content) != 0:
+            actions.user.display_response(content)
+
+    else:
+        notify("GPT Failure: Check API Key, Model, or Prompt")
+        print(response.json())
+
+def process_function_calls(insert_response, response):
+    try:
+        tool_calls = response.json()["choices"][0]["message"]["tool_calls"]
+        while tool_calls:
+            tool = tool_calls.pop()
+            first_argument = tool['function']['arguments']
+            try:
+                first_argument = json.loads(first_argument)['str']
+            except Exception as e:
+                notify(f"Argument Jason was malformed: {e}")
+            match tool['function']['name']:
+                case 'insert':
+                    insert_response(first_argument)
+                case 'display':
+                    actions.user.display_response(first_argument)
+                case 'notify':
+                    actions.user.notify_user(first_argument)
+                case 'search_for_command':
+                    actions.user.search_for_command(first_argument)
+    except Exception as e:
+        notify(f"No tool_calls found in response from LLM: {e}")
+
+@mod.action_class
+class UserActions:
+    def gpt_go(utterance: str, selected_text: str) -> str:
+        """Answer an arbitrary question"""
```

Comment:
I have mixed feelings about "go" as a verb, but I'm not strongly opposed. If we do go with such a vague verb, then we really need a good docstring here

Reply:
Updated the doc string

Reply:
I'm not so sure about the choice of the word go myself, but I don't see this as intentionally different from `model please` for example. The user is asking the model to do something on their behalf, and they don't really care how it is done. The function calling is an implementation detail that simply makes the LLM more aware of the impact it's having on the user's system. For that reason, the LLM can more intentionally switch between inserting text, displaying text, running commands, and sending notifications. I like having something that is ambiguous and quick since the function calling is not the point.

Reply:
Just as a thought, we obviously don't need to use model before `go`. I personally do not really like `go`. Something like `hey talon` may be interesting since it mimics the behavior of `hey siri` and more clearly signifies that there is some background guesswork

Reply:
Fwiw - I don't mind putting this behind the beta tag, if that is preferable.

Reply:
I asked ChatGPT for some suggestions and this is what it came up with:

🤖
Command Initiation 'Begin', 'Start', 'Commence'
🙏
Politeness 'Please', 'Kindly', 'Would you mind'
🗣️
Request 'Can you', 'Could you', 'Would you be able to'
🔍
Search 'Find', 'Search for', 'Look up'
📝
Writing 'Write', 'Compose', 'Draft'
📅
Scheduling 'Schedule', 'Set up', 'Arrange'
📧
Email 'Send an email', 'Compose an email', 'Draft an email'
📊
Analysis 'Analyze', 'Assess', 'Evaluate'
📌
Specificity 'Specifically', 'In particular', 'Regarding'

Reply:
I kind of like `model can you`; it's longer than `go` but pretty easy to say.

Reply:
`model can you` sounds good with me. It does start to sound similar to `model please`  and I still do like something like `hey $X`, but I don't want to be fastidious, so this is fine. 
We will just put it behind the beta tag until we align with user feedback

Reply:
I actually really like the command "model please" for this one. Tbh I would be tempted to just have this new function-calling command replace the existing "model please" if you have the tag activated

Alternatives:

- `"model act"`
- `"model do"`
- `"model fly"`
- `"model auto"`
- `"model run"`

Reply:
I'm also a fan of "please"; is there an option that could allow @C-Loftus to keep the existing behavior if he wants it?

Reply:
Maybe the play is to make a separate tag for function calling to replace "please"?

Reply:
Changed to `please` since I changed `can you` to be a superset of please.